src/extractor/edge_based_graph_factory.cpp

```diff
@@ -368,181 +375,237 @@ void EdgeBasedGraphFactory::GenerateEdgeExpandedEdges(
         util::Percent progress(log, m_node_based_graph->GetNumberOfNodes());
         // going over all nodes (which form the center of an intersection), we compute all
         // possible turns along these intersections.
-        for (const auto node_at_center_of_intersection :
-             util::irange(0u, m_node_based_graph->GetNumberOfNodes()))
-        {
-            progress.PrintStatus(node_at_center_of_intersection);
-
-            const auto shape_result =
-                turn_analysis.ComputeIntersectionShapes(node_at_center_of_intersection);
-
-            // all nodes in the graph are connected in both directions. We check all outgoing nodes
-            // to
-            // find the incoming edge. This is a larger search overhead, but the cost we need to pay
-            // to
-            // generate edges here is worth the additional search overhead.
-            //
-            // a -> b <-> c
-            //      |
-            //      v
-            //      d
-            //
-            // will have:
-            // a: b,rev=0
-            // b: a,rev=1 c,rev=0 d,rev=0
-            // c: b,rev=0
-            //
-            // From the flags alone, we cannot determine which nodes are connected to `b` by an
-            // outgoing
-            // edge. Therefore, we have to search all connected edges for edges entering `b`
-            for (const EdgeID outgoing_edge :
-                 m_node_based_graph->GetAdjacentEdgeRange(node_at_center_of_intersection))
-            {
-                const NodeID node_along_road_entering =
-                    m_node_based_graph->GetTarget(outgoing_edge);
+        std::mutex merge_mutex; // To control merging of all the data into the main data structures
+        std::uint64_t nodes_completed = 0;
+
+        tbb::parallel_for(
```

Comment:
After using the `tbb::parallel_pipeline` in #4137 I think it would be very easy to adapt it for this use case. The biggest advantage over using that over a manual `parallel_for` is that we can force processing to be in-order. That would remove the non-determinism we introduced here.